Add "Trino's summer of grammar" blog post for Trino 482 (#840)
trinodb/trino.io: _posts/2026-06-26-summer-of-grammar.md

---
layout: post
title: "Trino's summer of grammar"
author: "Martin Traverso, Mateusz Gajewski"
excerpt_separator: <!--more-->
image: /assets/blog/summer-of-grammar.jpeg
---

> A query engine is, before anything else, a language. And like any language, SQL

Reviewer: did you mean a "language processor"?

Author: reworded

is defined by its grammar: the predicates, operators, and forms you're allowed
to write down. Trino has always spoken SQL fluently, but the [ISO 9075
standard](https://www.iso.org/standard/76583.html) is a big book, and there have
always been a few corners of it we hadn't gotten around to implementing yet.

Trino 482 closes a remarkable number of those gaps in a single release. So many,

> in fact, that we started calling it the summer of grammar. This post walks

Reviewer: ```suggestion
in fact, that we started calling it the summer of grammar release. This post walks
```

Author: I would keep this as is. The previous sentence already says "in a single release," so repeating "release" here reads a bit redundant, and "calling it the summer of grammar" works as a nod to the title.

convincing as running it, every example below is live. Hit **Run** and watch
Trino 482 evaluate it for real.

<!--more-->

<script src="https://trysql.io/embed.js" async></script>

<style>
try-sql { display: block; margin: 1.5rem auto; }
try-sql iframe { display: block; margin: 0 auto; }
</style>

## `BETWEEN`, both ways

Let's start with an old friend. Everyone knows
[`x BETWEEN a AND b`](/docs/current/functions/comparison.html): it's just
shorthand for `a <= x AND x <= b`. The catch is that the order matters. If you
get the bounds backwards, the predicate is silently always false, because nothing
is simultaneously `>= 10` and `<= 1`.

The SQL standard has a fix for this that Trino didn't previously support: the
`SYMMETRIC` keyword. `x BETWEEN SYMMETRIC a AND b` treats the two bounds as an
unordered pair, so it's true whenever `x` falls between the smaller and the
larger, regardless of which you wrote first. `ASYMMETRIC` (the default) spells
out the classic order-sensitive behavior.

<try-sql version="482" height="200">
<pre data-query>SELECT 5 BETWEEN SYMMETRIC 10 AND 1 AS symmetric,
       5 BETWEEN ASYMMETRIC 10 AND 1 AS asymmetric</pre>
</try-sql>

> This is genuinely useful when the bounds come from columns or parameters and you

Reviewer: Should we mention that this also works for ranges that go from a negative value to a positive value?

Author: Negative-to-positive ranges aren't actually a special case here — plain `BETWEEN` already handles them fine (`x BETWEEN -5 AND 5` works as expected). `SYMMETRIC` only matters for the *order* of the bounds, i.e. when the lower one is written second (`-5 BETWEEN 5 AND -5`). So sign doesn't enter into it, and I'd rather not imply it does. I think the current example covers the distinction; happy to add a half-sentence stressing "it's purely about which bound you write first" if you think it needs underlining.

can't guarantee which one is larger.

## Three-valued logic, made explicit

No discussion of SQL is complete without an exploration of the semantics of
`null`. SQL uses [three-valued
logic](https://en.wikipedia.org/wiki/Three-valued_logic#Application_in_SQL): a
boolean expression can be `true`, `false`, or *unknown* (represented by `null`).
That third value is where a lot of subtle bugs live, because `NOT (a > b)` is
*not* the same as `a <= b` once `null` enters the picture.

The standard's answer is the
[`IS [NOT] TRUE`, `IS [NOT] FALSE`, and `IS [NOT]
UNKNOWN`](/docs/current/functions/comparison.html) predicates, and they now work

> in Trino. Unlike `=`, these always return `true` or `false` (never `null`), which

Reviewer: ```suggestion
in Trino. Unlike `=`, these always return `true` or `false`, and never `null`. That 
```

is exactly what you want when you need to collapse three-valued logic back down
to two.

<try-sql version="482" height="220">
<pre data-query>SELECT (1 &gt; 2) IS FALSE AS is_false,
       (1 &lt; 2) IS TRUE AS is_true,
       CAST(NULL AS boolean) IS UNKNOWN AS is_unknown</pre>
</try-sql>

## Looking inside subqueries

Two new predicates let you ask questions about the *shape* of a subquery's
results, not just its values.

> The `UNIQUE` predicate is true when no two rows returned by a subquery are equal.

Reviewer: maybe 

```suggestion
The `UNIQUE` predicate is `true` when no two rows returned by a subquery are equal.
```

It's the declarative way to assert "this subquery has no duplicates" without
contorting yourself into a `GROUP BY ... HAVING count(*) > 1` and checking
whether it's empty.

<try-sql version="482" height="200">
<pre data-query>SELECT UNIQUE (SELECT x FROM (VALUES 1, 2, 3) t(x)) AS all_distinct,
       UNIQUE (SELECT x FROM (VALUES 1, 2, 2) t(x)) AS has_duplicate</pre>
</try-sql>

The `MATCH` predicate tests whether a row value appears in a subquery's results.

> Add the `UNIQUE` keyword and it's true only when the row matches *exactly one*

Reviewer: maybe again 

```suggestion
Add the `UNIQUE` keyword and it's `true` only when the row matches *exactly one*
```

row, a neat way to express "this value exists, and there's only one of it."

<try-sql version="482" height="200">
<pre data-query>SELECT 2 MATCH (SELECT x FROM (VALUES 1, 2, 3) t(x)) AS found,
       2 MATCH UNIQUE (SELECT x FROM (VALUES 1, 2, 2) t(x)) AS found_once</pre>
</try-sql>

> ## `CASE` gets some opinions

Reviewer: Did you mean 

```suggestion
## `CASE` gets some options
```

Author: "opinions" is intentional here. The joke is that a simple `CASE` could only test equality before, and now it can weigh ranges and other predicates, so it has gained a bit of judgment. Keeping it.

branch.

No longer. The `WHEN` clauses of a simple `CASE` can now contain comparison
operators, `BETWEEN`, and `IS NULL`, so you write the operand once and let each
branch apply its own predicate to it.

<try-sql version="482" height="240">
<pre data-query>SELECT x,
       CASE x
           WHEN &lt; 0 THEN 'negative'
           WHEN BETWEEN 0 AND 9 THEN 'small'
           ELSE 'large'
       END AS bucket
FROM (VALUES -5, 3, 100) t(x)</pre>
</try-sql>

## Time, locally

Trino has long supported
[`AT TIME ZONE`](/docs/current/functions/datetime.html) to render a timestamp in
a specific zone. The standard also defines `AT LOCAL`, which converts a value to
the session's own time zone without you having to name it explicitly. It's the
difference between "show me this in `America/Los_Angeles`" and "show me this
wherever I happen to be."

<try-sql version="482" height="200">
<pre data-query>SELECT TIMESTAMP '2026-06-21 14:00:00 America/Los_Angeles' AT LOCAL AS in_my_zone</pre>
</try-sql>

The result above is rendered in the session's time zone. Change the session zone
and the same expression follows you there.

> ## Calling functions by name

Reviewer: ```suggestion
## Calling functions with named parameters
```

or named arguments

When a function takes more than two or three arguments, positional calls become a
guessing game: which argument was the fourth one, again? Trino 482 adds the
standard `name => value` syntax for passing arguments by name, in any order.

This is especially handy for [table
functions](/docs/current/functions/table.html), which often have several optional
parameters, but it works for any function whose parameters are named, including
the [user-defined functions](/docs/current/udf.html) you write yourself. Notice

> that the call below supplies the arguments in the *opposite* order from the

Reviewer: ```suggestion
that the following call supplies the arguments in the *opposite* order from the
```

Author: "below" works here since the runnable example is right below this line, so keeping as is.

</try-sql>

There's a new [`ends_with`](/docs/current/functions/string.html) function, the
obvious companion to the long-standing `starts_with`, and a
[`ROW::fields`](/docs/current/functions/row.html) function that returns the field
names of a `row` value, which is handy when you're working with anonymous or
programmatically-built rows.

<try-sql version="482" height="220">
<pre data-query>SELECT ends_with('trino.io', '.io') AS yes,
       ROW::fields(CAST(ROW(1, 'a') AS ROW(id integer, name varchar))) AS field_names</pre>
</try-sql>

Perhaps the most fun addition is ergonomic rather than functional: you can now
invoke string functions as *methods* on character values. `'Trino'.length()` is
just another way to write `length('Trino')`, and the `type::function` form lets
you reach a function through its type. It reads naturally when you're chaining
transformations.

<try-sql version="482" height="200">
<pre data-query>SELECT 'Trino'.length() AS length,
       varchar::chr(65) AS letter_a</pre>
</try-sql>

## `char` and `varchar` make peace

This is the one to read carefully, because it's a deliberate **breaking change**.

> For historical reasons, Trino used to implicitly coerce `varchar` to `char`,

Reviewer: Maybe clarify that the new way it works is aligned with the spec (i hope)

Author: I would rather not claim spec alignment here, because it is not quite true. The ISO default `CHAR` collation is PAD SPACE, so under the strict spec `'abc' = 'abc   '` is true (the shorter value gets blank-padded). The new behavior treats char-to-varchar comparison as NO PAD, so that comparison is now false. It matches `varchar` / NO PAD semantics and what most engines and users expect, but it is a departure from strict ISO `CHAR` padding rather than a move toward it (char-to-char comparisons stay blank-padded). So I think an "aligned with the spec" note would be misleading and I would leave it out.

which dragged in `char`'s blank-padded comparison semantics and surprised just

> about everyone. Trino 482 reverses the direction: a

Reviewer: I kinda recall we had some docs around that .. I assume you updated those

Author: yes: https://trino.io/docs/current/language/types.html#char-and-varchar-coercion-and-comparison

inside a JSON document with [`json_exists`](/docs/current/functions/json.html)
using a regular expression rather than exact matches:

<try-sql version="482" height="200">
<pre data-query>SELECT json_exists('["foobar", "baz"]', 'lax $[*] ? (@ like_regex "^foo")') AS has_match</pre>
</try-sql>

And `row` and `array` values that contain `null` elements can now be compared and
ordered (in `ORDER BY`, `DISTINCT`, `min`, `max`, and range comparisons) where
they previously would have failed. `null` elements sort consistently to the end:

<try-sql version="482" height="220">
<pre data-query>SELECT x
FROM (VALUES (ARRAY[1, 1]), (ARRAY[1, 2]), (ARRAY[1, NULL])) t(x)
ORDER BY x</pre>
</try-sql>

## Wrapping up

None of these features is, on its own, the headline of a release. But taken
together they make Trino's dialect of SQL noticeably more complete and more
pleasant to write: fewer workarounds, fewer "why doesn't the standard form work
here," and a few genuinely new tools. That's a good summer's work on the grammar.

For the complete list, including the connector and engine improvements we didn't
cover here, see the [Trino 482 release

> notes](/docs/current/release/release-482.html). Happy querying!

Reviewer: ```suggestion
notes](/docs/current/release/release-482.html). 

Happy querying!
```
